### PR TITLE
[API change] Pooling backward solvers (except GetWorkspaceSize)

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
This PR contains both backward pooling solvers.
It now includes bwd transposed solvers.

### CODE REVIEW

**Comment by junliume** [comment]
@DrizztDoUrden
> [2021-11-19T21:50:02.144Z] 50/92 Test #39: test_pooling3d ........................................***Failed  Error regular expression found in output. Regex=[FAILED]271.25 sec
[2021-11-19T21:50:02.144Z] /var/jenkins/workspace/s_MIOpen_ddu_pooling-bwd-solvers/build/bin/test_pooling3d --float --index_type miopenIndexUint32 --mode miopenPoolingMax --verify_indices 0 --input 16 64 3 4 4 --lens 2 2 2 --strides 2 2 2 --pads 0 0 0 --wsidx 1 
[2021-11-19T21:50:02.144Z] **FAILED: /var/jenkins/workspace/s_MIOpen_ddu_pooling-bwd-solvers/src/ocl_kernel.cpp:87: Running kernel failed:  Invalid Kernel Args**

**Comment by atamazov** [comment]
@DrizztDoUrden Is testing done? Description says "in process"

**Comment by atamazov** [comment]
@zjing14 It would be nice to get your blessing before merging this.

**Review by atamazov — DISMISSED** [review]
LGTM

**Comment by atamazov** [comment]
@averinevg If you have time, could you please review? I would like to double-check if this refactoring PR is really clean and safe (No Functional Changes).

**Comment by zjing14** [comment]
LGTM

File: src/solver/pooling/backward2d.cpp
Comment (averinevg) [inline src/solver/pooling/backward2d.cpp:99]: It's better to add a comment to clarify what's going on here

Comment (atamazov) [reply src/solver/pooling/backward2d.cpp:99]: @averinevg This is ascii art, I guess ;)

Seriously, this code existed before this PR. This change simply moves it to another place. You can ask the original author of this code to add comments (you need to open a separate ticket for that).

Comment (DrizztDoUrden) [reply src/solver/pooling/backward2d.cpp:99]: I have also formatted such pieces to be less of an ascii art and more of an understandable code. But otherwise, it has only been moved.

Comment (atamazov) [reply src/solver/pooling/backward2d.cpp:99]: @DrizztDoUrden Please do not destroy art for the sake of utility! :rofl:

Comment (junliume) [reply src/solver/pooling/backward2d.cpp:99]: 🤣 any improvements to readability are definitely welcomed!

**Comment by atamazov** [comment]
@DrizztDoUrden Can you please fix merge conflicts, thanks!

**Comment by junliume** [comment]
There are some CI issues when latest develop is merged:
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/ddu%2Fpooling-bwd-solvers/9/pipeline

```
[2022-01-04T23:29:23.889Z] In file included from /var/jenkins/workspace/s_MIOpen_ddu_pooling-bwd-solvers/src/solver/pooling/forward2d.cpp:27:

[2022-01-04T23:29:23.889Z] /var/jenkins/workspace/s_MIOpen_ddu_pooling-bwd-solvers/src/include/miopen/pooling/solvers.hpp:90:35: error: unknown template name 'SolverBase'

[2022-01-04T23:29:23.889Z] struct PoolingBackward2d : public SolverBase<OldStyleProblemDescription>

[2022-01-04T23:29:23.889Z]                                   ^

**Comment by averinevg** [comment]
@DrizztDoUrden @junliume 
> ... error: unknown template name 'SolverBase'

I would fix it

**Review by junliume — APPROVED** [review]

**Comment by junliume** [comment]
@DrizztDoUrden could you merge develop into this branch and resolve the conflicts? Thanks!

**Comment by atamazov** [comment]
@DrizztDoUrden 
image

Also it may worth updating from develop.

**Comment by DrizztDoUrden** [comment]
@junliume I have a problem with the API here. Backward pooling call has a constant workspace pointer. That doesn't allow us to use it for the transpose. But making it non-const shouldn't be a breaking change. Should I update the API?

**Comment by junliume** [comment]
> @junliume I have a problem with the API here. Backward pooling call has a constant workspace pointer. That doesn't allow us to use it for the transpose. But making it non-const shouldn't be a breaking change. Should I update the API?

@DrizztDoUrden sorry for the late reply. I would vote for updating the API as long as it is not breaking.
@averinevg @atamazov @JehandadKhan @ce1adon @zjing14  @asroy  what do you think?

**Comment by averinevg** [comment]
> > @junliume I have a problem with the API here. Backward pooling call has a constant workspace pointer. That doesn't allow us to use it for the transpose. But making it non-const shouldn't be a breaking change. Should I update the API?
> 
> @DrizztDoUrden sorry for the late reply. I would vote for updating the API as long as it is not breaking. @averinevg @atamazov @JehandadKhan @ce1adon @zjing14 @asroy what do you think?

@DrizztDoUrden Could you please show how the new API will look like compared to the old one?

**Comment by DrizztDoUrden** [comment]
new:
```
MIOPEN_EXPORT miopenStatus_t miopenPoolingBackward(miopenHandle_t handle,
                                                   const miopenPoolingDescriptor_t poolDesc,
                                                   const void* alpha,
                                                   const miopenTensorDescriptor_t yDesc,
                                                   const void* y,
                                                   const miopenTensorDescriptor_t dyDesc,
                                                   const void* dy,
                                                   const miopenTensorDescriptor_t xDesc,
                                                   const void* x,
                                                   const void* beta,
                                                   const miopenTensorDescriptor_t dxDesc,
                                                   void* dx,
                                                   void* workSpace);
```

old:
```
MIOPEN_EXPORT miopenStatus_t miopenPoolingBackward(miopenHandle_t handle,
                                                   const miopenPoolingDescriptor_t poolDesc,
                                                   const void* alpha,
                                                   const miopenTensorDescriptor_t yDesc,
                                                   const void* y,
                                                   const miopenTensorDescriptor_t dyDesc,
                                                   const void* dy,
                                                   const miopenTensorDescriptor_t xDesc,
                                                   const void* x,
                                                   const void* beta,
                                                   const miopenTensorDescriptor_t dxDesc,
                                                   void* dx,
                                                   const void* workSpace);
```

**Comment by atamazov** [comment]
> > @junliume I have a problem with the API here. Backward pooling call has a constant workspace pointer. That doesn't allow us to use it for the transpose. But making it non-const shouldn't be a breaking change. Should I update the API?
> 
> @averinevg @atamazov @JehandadKhan @ce1adon @zjing14 @asroy what do you think?

No objections. Actually we do not have anything better than removal of constness, if we want to engage transpose to implement NHWC pooling quickly.

:warning: This may lead to build errors of the user apps, if they actually pass const workspace ptr to `miopenPoolingBackward()`. But the fix of such user apps should be trivial.

**Comment by DrizztDoUrden** [comment]
They would have to create a new const buffer and to copy data from the forward workspace (which may not be const). But yes, I have somehow missed that it is actually kind of breaking in theory.

**Comment by junliume** [comment]
@DrizztDoUrden we have some HIP tidy issues: [http://micimaster.amd.com/blue/rest/organizations/jenkins/pipelines/MLLibs/pipelines/MIOpen/branches/ddu%252Fpooling-bwd-solvers/runs/21/nodes/13/log/?start=0](http://micimaster.amd.com/blue/rest/organizations/jenkins/pipelines/MLLibs/pipelines/MIOpen/branches/ddu%252Fpooling-bwd-solvers/runs/21/nodes/13/log/?start=0)

**Comment by DrizztDoUrden** [comment]
Updated the API

**Review by junliume — APPROVED** [review]

**Comment by atamazov** [comment]
I recommend marking this PR with `"[API change]"` in order to make this commit immediately visible in the history.

File: src/solver/pooling/backward2d.cpp
Comment (atamazov) [inline src/solver/pooling/backward2d.cpp:123]: @DrizztDoUrden It seems like we need to swap `_h` and `_w` here to fix 2d pooling.

**Pull request merged**
